sensor/extrinsic_map_based_calibrator/src/extrinsic_map_based_preprocessing.cpp

```diff
@@ -46,52 +47,133 @@ PointCloudT::Ptr ExtrinsicMapBasedPreprocessing::preprocessing(
-  seg.setRadiusLimits(config_.ransac_config.min_radius, std::numeric_limits<double>::max());
-  seg.setMethodType(pcl::SAC_RANSAC);
-  seg.setDistanceThreshold(config_.ransac_config.distance_threshold);
-  seg.setInputCloud(pcl_sensor);
-  seg.setMaxIterations(config_.ransac_config.max_iteration);
-  seg.setModelType(pcl::SACMODEL_PLANE);
-  seg.segment(*output_inliers, *output_coefficients);
+  PointCloudT::Ptr & pcl_filtered_sensor,
+  PointCloudT::Ptr & inliers_pointcloud_pcl) const
 
-  // extract target cloud and floor cloud
-  PointCloudT::Ptr cloud_floor(new PointCloudT);
+{
   PointCloudT::Ptr cloud_target(new PointCloudT);
-  pcl::ExtractIndices<pcl::PointXYZ> extract;
-  extract.setInputCloud(pcl_sensor);
-  extract.setIndices(output_inliers);
-  extract.setNegative(false);
-  extract.filter(*cloud_floor);
-  extract.setNegative(true);
-  extract.filter(*cloud_target);
+  Eigen::Vector4d ground_plane_model;
+  extractGroundPlane(pcl_sensor, ground_plane_model, inliers_pointcloud_pcl/* , cloud_target */);
+  PointCloudT::Ptr cloud_ground_plane(new PointCloudT);
+  separatePointCloud(pcl_sensor, ground_plane_model, cloud_ground_plane, cloud_target);
 
   // voxel grid filtering
   PointCloudT::Ptr cloud_voxel_filtered(new PointCloudT);
   pcl::VoxelGrid<pcl::PointXYZ> vg;
-  vg.setInputCloud(pcl::make_shared<PointCloudT>(*cloud_floor));
+  vg.setInputCloud(boost::make_shared<PointCloudT>(*cloud_ground_plane));
   vg.setLeafSize(config_.ransac_config.voxel_grid_size, config_.ransac_config.voxel_grid_size, config_.ransac_config.voxel_grid_size);
   vg.filter(*cloud_voxel_filtered);
 
+  pcl::VoxelGrid<pcl::PointXYZ> vg_target;
+  PointCloudT::Ptr filtered_cloud_target(new PointCloudT);
+  vg_target.setInputCloud(boost::make_shared<PointCloudT>(*cloud_target));
+  vg_target.setLeafSize(config_.ransac_config.voxel_grid_size/2, config_.ransac_config.voxel_grid_size/2, config_.ransac_config.voxel_grid_size/2);
+  vg_target.filter(*filtered_cloud_target);
+
   // merge target cloud and downsampled floor cloud
-  *pcl_filtered_sensor = *cloud_voxel_filtered + *cloud_target;
+  *pcl_filtered_sensor = *cloud_voxel_filtered + *filtered_cloud_target;
 
 }
 
+bool ExtrinsicMapBasedPreprocessing::extractGroundPlane(
+  const PointCloudT::Ptr & pointcloud, Eigen::Vector4d & model,
+  PointCloudT::Ptr & plane_pointcloud/* , PointCloudT::Ptr & target_pointcloud */) const
+{
+  std::vector<pcl::ModelCoefficients> models;
+
+  // Obtain an idea of the ground plane using PCA
+  // under the asumption that the axis with less variance will be the ground plane normal
+  pcl::PCA<pcl::PointXYZ> pca;
+  pca.setInputCloud(pointcloud);
+  Eigen::MatrixXf vectors = pca.getEigenVectors();
+  Eigen::Vector3f rough_normal = vectors.col(2);
+
+  // Use RANSAC iteratively until we find the ground plane
+  // Since walls can have more points, we filter using the PCA-based hypothesis
+  pcl::ModelCoefficients::Ptr coefficients(new pcl::ModelCoefficients);
```

Comment:
I discussed way back with you but please do not use this idea on the XX1.
The while idea and processing requires an ideal environment that is not the one you get normally when using this calibration